x/evm/keeper/precompiles.go

```diff
@@ -148,83 +147,77 @@ func (k *Keeper) WithPrecompiles(precompiles map[common.Address]vm.PrecompiledCo
 	return k
 }
 
-// Precompiles returns the subset of the available precompiled contracts that
-// are active given the current parameters.
-func (k Keeper) Precompiles(
+// GetCachedPrecompiles returns the subset of the available precompiled contracts that
+// are initalized in memory. If a precompile is not found in the cache, it will be
+// instantiated and added to the cache.
+func (k Keeper) GetCachedPrecompiles(
+	ctx sdk.Context,
 	activePrecompiles ...common.Address,
 ) map[common.Address]vm.PrecompiledContract {
-	activePrecompileMap := make(map[common.Address]vm.PrecompiledContract)
-
+	newActivePrecompileMap := make(map[common.Address]vm.PrecompiledContract)
 	for _, address := range activePrecompiles {
-		precompile, ok := k.precompiles[address]
+		// Check cached precompiles
+		cachedPrecompile, ok := k.precompiles[address]
+
+		// If precompile is not found on cache, try to initiate it
+		// It can only be an erc20 precompile
 		if !ok {
-			panic(fmt.Sprintf("precompiled contract not initialized: %s", address))
+			precompile, err := k.erc20Keeper.InstantiateERC20Precompile(ctx, address)
+			if err != nil {
+				panic(fmt.Sprintf("precompiled contract not initialized: %s", address))
+			}
+			fmt.Println(precompile.Address())
```

Comment:
if we are doing this during runtime we probably also do not want to `panic` anywhere in the code right?